odoo/addons/base/tests/test_ir_mail_server_smtpd.py

```diff
@@ -0,0 +1,460 @@
+        })
+
+        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+        ssl_context.load_cert_chain(self.ssl_server.cert, self.ssl_server.key)
+
+        matrix = [
+            # client, server, error_pattern
+            ('none', 'ssl',
+                r"The server has closed the connection unexpectedly\. "
+                r"Check configuration served on this port number\.\n "
+                r"Connection unexpectedly closed: timed out"),
+            ('none', 'starttls',
+                r"The server refused the sender address \(noreply@localhost\) with error "
+                r"b'Must issue a STARTTLS command first'"),
+            ('starttls', 'none',
+                r"An option is not supported by the server:\n "
+                r"STARTTLS extension not supported by server\."),
+            ('starttls', 'ssl',
+                r"The server has closed the connection unexpectedly\. "
+                r"Check configuration served on this port number\.\n "
+                r"Connection unexpectedly closed: timed out"),
+            ('ssl', 'none',
+                r"An SSL exception occurred\. "
+                r"Check connection security type\.\n "
+                r".*?wrong version number"),
+            ('ssl', 'starttls',
+                r"An SSL exception occurred\. "
+                r"Check connection security type\.\n "
+                r".*?wrong version number"),
+        ]
+
+        for client_encryption, server_encryption, error_pattern in matrix:
+            with self.subTest(server_encryption=server_encryption,
+                              client_encryption=client_encryption):
+                mail_server.smtp_encryption = client_encryption
+                with self.start_smtpd(server_encryption, ssl_context, auth_required=False):
+                    with self.assertRaises(UserError) as capture:
+                        mail_server.test_smtp_connection()
+                    self.assertRegex(capture.exception.args[0], error_pattern)
+
+    def test_man_in_the_middle_matrix(self):
+        """
+        Simulate that a pirate was successful at intercepting the live
+        traffic in between the Odoo server and the legitimate SMTP
+        server.
+        """
+        mail_server = self.env['ir.mail_server'].create({
+            'name': 'test smtpd',
+            'from_filter': 'localhost',
+            'smtp_host': 'localhost',
+            'smtp_port': self.port,
+            'smtp_authentication': 'login',
+            'smtp_user': self.user_demo.email,
+            'smtp_pass': PASSWORD,
+            'smtp_ssl_certificate': b64encode(self.ssl_client.cert.read_bytes()),
+            'smtp_ssl_private_key': b64encode(self.ssl_client.key.read_bytes()),
+        })
+
+        cert_good = self.ssl_server
+        cert_bad = self.ssl_self_signed
+        host_good = 'localhost'
+        host_bad = 'notlocalhost'
+
+        # for now it doesn't raise any error for bad cert/host
```

Comment:
Side note: could be explained in commit message that we somehow hope to fix that in a near future